src/game/assets.rs

```diff
@@ -44,58 +50,78 @@ pub enum SfxAsset {
     Step4,
 }
 
-#[derive(Resource, Reflect, Deref, DerefMut)]
-pub struct SfxAssets(HashMap<SfxAsset, Handle<AudioSource>>);
-
-impl SfxAssets {
-    pub fn new(asset_server: &AssetServer) -> Self {
-        let mut assets = HashMap::new();
-
-        assets.insert(
-            SfxAsset::ButtonHover,
-            asset_server.load("audio/sfx/button_hover.ogg"),
-        );
-        assets.insert(
-            SfxAsset::ButtonPress,
-            asset_server.load("audio/sfx/button_press.ogg"),
-        );
-        assets.insert(SfxAsset::Step1, asset_server.load("audio/sfx/step1.ogg"));
-        assets.insert(SfxAsset::Step2, asset_server.load("audio/sfx/step2.ogg"));
-        assets.insert(SfxAsset::Step3, asset_server.load("audio/sfx/step3.ogg"));
-        assets.insert(SfxAsset::Step4, asset_server.load("audio/sfx/step4.ogg"));
-
-        Self(assets)
-    }
+impl AssetKey for SfxKey {
+    type Asset = AudioSource;
+}
 
-    pub fn all_loaded(&self, assets: &Assets<AudioSource>) -> bool {
-        self.0.iter().all(|(_, handle)| assets.contains(handle))
+impl FromWorld for AssetMap<SfxKey> {
+    fn from_world(world: &mut World) -> Self {
+        let asset_server = world.resource::<AssetServer>();
+        [
+            (
+                SfxKey::ButtonHover,
+                asset_server.load("audio/sfx/button_hover.ogg"),
+            ),
+            (
+                SfxKey::ButtonPress,
+                asset_server.load("audio/sfx/button_press.ogg"),
+            ),
+            (SfxKey::Step1, asset_server.load("audio/sfx/step1.ogg")),
+            (SfxKey::Step2, asset_server.load("audio/sfx/step2.ogg")),
+            (SfxKey::Step3, asset_server.load("audio/sfx/step3.ogg")),
+            (SfxKey::Step4, asset_server.load("audio/sfx/step4.ogg")),
+        ]
+        .into()
     }
 }
 
 #[derive(PartialEq, Eq, Hash, Reflect)]
-pub enum SoundtrackAsset {
+pub enum SoundtrackKey {
     Credits,
     Gameplay,
 }
 
+impl AssetKey for SoundtrackKey {
+    type Asset = AudioSource;
+}
+
+impl FromWorld for AssetMap<SoundtrackKey> {
+    fn from_world(world: &mut World) -> Self {
+        let asset_server = world.resource::<AssetServer>();
+        [
+            (
+                SoundtrackKey::Credits,
+                asset_server.load("audio/soundtracks/Monkeys Spinning Monkeys.ogg"),
+            ),
+            (
+                SoundtrackKey::Gameplay,
+                asset_server.load("audio/soundtracks/Fluffing A Duck.ogg"),
+            ),
+        ]
+        .into()
+    }
+}
+
+pub trait AssetKey: Sized {
```

Comment:
I think this trait is no longer needed

Reply:
It's needed so that `AssetMap` (or `HandleMap`) only requires one generic.